src/groups/mqb/mqbi/mqbi_dispatcher.cpp

```diff
@@ -284,11 +288,133 @@ DispatcherReceiptEvent::~DispatcherReceiptEvent()
 // class DispatcherEvent
 // ---------------------
 
+DispatcherEvent::DispatcherEvent(bslma::Allocator* allocator)
+: d_type(DispatcherEventType::e_UNDEFINED)
+, d_source_p(0)
+, d_destination_p(0)
+, d_ackMessage()
+, d_blob_sp(0, allocator)
+, d_options_sp(0, allocator)
+, d_clusterNode_p(0)
+, d_confirmMessage()
+, d_rejectMessage()
+, d_controlMessage(allocator)
+, d_guid(bmqt::MessageGUID())
+, d_isRelay(false)
+, d_partitionId(mqbi::Storage::k_INVALID_PARTITION_ID)
+, d_putHeader()
+, d_queueHandle_p(0)
+, d_queueId(bmqp::QueueId::k_UNASSIGNED_QUEUE_ID)
+, d_subQueueInfos(allocator)
+, d_messagePropertiesInfo()
+, d_compressionAlgorithmType(bmqt::CompressionAlgorithmType::e_NONE)
+, d_isOutOfOrder(false)
+, d_genCount(0)
+, d_callback(allocator)
+, d_finalizeCallback(allocator)
+{
+    // NOTHING
+}
+
 DispatcherEvent::~DispatcherEvent()
 {
     // NOTHING (interface)
 }
 
+void DispatcherEvent::reset()
+{
+    switch (d_type) {
+    case mqbi::DispatcherEventType::e_UNDEFINED: {
+        // NOTHING
+    } break;
+    case mqbi::DispatcherEventType::e_DISPATCHER: {
+        if (!d_finalizeCallback.empty()) {
+            // We only set finalizeCallback on e_DISPATCHER events
+
+            // TODO(678098): make a special event type that handles this case
+            d_finalizeCallback();
+        }
+
+        d_callback.reset();
+        d_finalizeCallback.reset();
+    } break;
+    case mqbi::DispatcherEventType::e_CALLBACK: {
+        d_callback.reset();
+    } break;
+    case mqbi::DispatcherEventType::e_CONTROL_MSG: {
+        d_controlMessage.reset();
+    } break;
+    case mqbi::DispatcherEventType::e_CONFIRM: {
+        d_blob_sp.reset();
+        d_clusterNode_p  = 0;
+        d_confirmMessage = bmqp::ConfirmMessage();
+        d_isRelay        = false;
+    } break;
+    case mqbi::DispatcherEventType::e_REJECT: {
+        d_blob_sp.reset();
+        d_clusterNode_p = 0;
+        d_rejectMessage = bmqp::RejectMessage();
+        d_isRelay       = false;
+        d_partitionId   = mqbi::Storage::k_INVALID_PARTITION_ID;
```

Comment:
Both `constructor`/`reset` are not inlined according to `perf`.
We can freely move them to `cpp` file without performance loss. This allow to include `mqbi_storage.h` in `cpp` and we don't need to expose this header in `mqbi_dispatcher.h`